Add Metrics

f2prateek opened the pull request:
Ref: https://paper.dropbox.com/doc/Analytics.js-Metrics-SDD-1hAD90lqGS4aZxHAHYPu7

This makes two main changes:
* Add a metrics client for analytics.js to report metrics to
* Integrate metrics client into analytics.js to report key metrics

The metrics client buffers metrics in memory and periodically uploads them to a server. By default, the metrics client does not upload any metrics (though it points to api.segment.io). It also automatically flushes the buffer when an error metric is recorded. Lastly - the client supports sampling so we can control the metrics being reported.

Metrics are invoked for the following operations:
* A metric is recorded for each method invoked in analytics.js.
* A metric is recorded for each attempt to invoke an analytics.js integration.
* A metric is recorded for each failed attempt when invoking an analytics.js integration.

calvinfo commented on lib/metrics.js:
```diff
@@ -0,0 +1,88 @@
+ * Increments the counter identified by name and tags by one.
+ *
+ * @param {String} metric Name of the metric to increment.
+ * @param {Array} tags Dimensions associated with the metric.
```
Minor nit, but we should make this `{Object}` ?

f2prateek replied:
Ah yes - we should.

calvinfo commented:
+1 to a max queue length, but overall this is looking good. nice work prateek

calvinfo commented on test/metrics.test.js:
```diff
@@ -0,0 +1,135 @@
+  var spy;
+
+  beforeEach(function() {
+    xhr = sinon.useFakeXMLHttpRequest();
```
is the reason the tests are failing that this isn't IE9 compatible?

https://caniuse.com/#feat=xhr2

f2prateek replied:
it could be! i'll look into it a bit more. Unfortunately, the IE tests have been flaky even before this change though

f2prateek commented:
@calvinfo @Peripheral1994 We didn't want to do it for v1 to keep the code simple, but can definitely add that!

When the max queue size is reached, we have two options, we can either:

* discard messages
* trigger a flush

If we do the first, it might skew our sampling a bit. If we do the latter we run the risk that we accidentally trigger too many flushes (e.g. if there was a bug causing too many metrics to be recorded).

Given those two tradeoffs and the purpose of our metrics, I'm leaning towards discarding messages when the queue is full for now, but let me know what you think!

calvinfo commented:
I am +1 on that!

f2prateek commented:
@Peripheral1994 @calvinfo Added a `maxQueueSize` option. Still looking into the IE errors a bit more.

calvinfo commented:
Sounds good, looking at that caniuse graph I'm 90% sure it's the XMLHTTPRequest that we're faking. Multiple people have reported that isn't compatible, and sinon might choke on parts of it due to the way we're doing the spys/mocking.

f2prateek commented:
@calvinfo yeah that's I came to conclude as well. Since the endpoint doesn't support JSONP, I added a check to skip making the request and updated tests accordingly.

f2prateek commented:
hey @Peripheral1994 @danielstjules @calvinfo, mind taking a final look here?
